### PR TITLE
feat: Add the http clients/dtos of KVS and registry for Keeper

### LINKED ISSUE #873: Add the http clients/dtos of KVS and registry for Keeper
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for core-keeper
https://github.com/edgexfoundry-holding/edgex-core-keeper/tree/edgex-go-v3.2.0-dev.10

### Description [**REQUIRED**]
 Add the http clients/dtos of KVS (key-value APIs) and registry for Keeper

### PR DESCRIPTION
**Pull request by jackchenjc**
closes #873 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

**Comment by cloudxxx8** [comment]
@jackchenjc may we move those clients to config and registry module respectively?

**Comment by cloudxxx8** [comment]
@lenny-intel we can dicuss in the tomorrow's meeting.
Since keeper is in the momo repo edgex-go, its dtos and models should be located in the go-mod-core-contracts.
The client lib is defined here for other go services.  There is no direct dependency from go-mod-registry and go-mod-configuration right now.

**Comment by cloudxxx8** [comment]
@lenny-intel according to the Core WG meeting discussion today, this PR can be reviewed and merged first

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: clients/http/kvs.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+// KVSClient is the REST client for invoking the key-value APIs(/kvs/*) from Core Keeper
+type KVSClient struct {
+	baseUrl      string
+	authInjector interfaces.AuthenticationInjector
+}
+
+// NewKVSClient creates an instance of KVSClient
+func NewKVSClient(baseUrl string, authInjector interfaces.AuthenticationInjector) interfaces.KVSClient {
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply clients/http/kvs.go:27]: Hi Lenny,

You're right, this client should be used in go-mod-configuration and we will encounter the issue you mentioned.

Would you please give us more hints about your suggestion, like what exactly should be added to the struct,  thanks.

File: dtos/requests/registration.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	Registration       dtos.Registration `json:"registration"`
+}
+
+func (a *AddRegistrationRequest) Validate() error {
+	err := a.Registration.Validate()
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/requests/registration.go:24]: They should have already been implemented [here](https://github.com/edgexfoundry/go-mod-core-contracts/pull/879/files#diff-424d0d6b55f61c7c1a8a6a6c6fc5d77d992b78f2e0bf995db6b4e08cd766cc48R32-R59).

[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/requests/registration.go:24]: The conversion functions to and from model are currently declared in [dto/registration.go](https://github.com/edgexfoundry/go-mod-core-contracts/pull/879/files#diff-424d0d6b55f61c7c1a8a6a6c6fc5d77d992b78f2e0bf995db6b4e08cd766cc48R32-R59), which is consistent with existing DTOs, would you suggest to move them to `dtos/requests/registration.go`?

File: clients/http/kvs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackchenjc) [reply clients/http/kvs.go:27]: Hi @lenny-intel,
When you have a moment, could you please review this conversation, thanks.

File: clients/interfaces/kvs.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"context"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/dtos/requests"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/dtos/responses"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/errors"
+)
+
+// KVSClient defines the interface for interactions with the kvs endpoint on the Edge Xpert core-keeper service.
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply clients/interfaces/kvs.go:16]: Done.

File: clients/interfaces/registry.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"context"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/dtos/requests"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/dtos/responses"
+	"github.com/edgexfoundry/go-mod-core-contracts/v3/errors"
+)
+
+// RegistryClient defines the interface for interactions with the registry endpoint on the Edge Xpert core-keeper service.
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply clients/interfaces/registry.go:16]: Done.

File: common/constants.go [hunk 5]
```diff
@@ -176,13 +182,18 @@ const (
...
 	Interval      = "interval"
+	Key           = "key"
+	ServiceId     = "serviceId"
 
 	Offset       = "offset"         //query string to specify the number of items to skip before starting to collect the result set.
 	Limit        = "limit"          //query string to specify the numbers of items to return
 	Labels       = "labels"         //query string to specify associated user-defined labels for querying a given object. More than one label may be specified via a comma-delimited list
 	PushEvent    = "ds-pushevent"   //query string to specify if an event should be pushed to the EdgeX system
 	ReturnEvent  = "ds-returnevent" //query string to specify if an event should be returned from device service
-	RegexCommand = "ds-regexcmd"    //query string to specify if the command name is in regular expression format
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply common/constants.go:185]: It was removed accidentally, already reverted.

File: dtos/registration.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	dto.Port = r.Port
+	dto.LastConnected = r.LastConnected
+	dto.HealthCheck.Type = r.HealthCheck.Type
+	dto.HealthCheck.Path = r.HealthCheck.Path
+	dto.HealthCheck.Interval = r.HealthCheck.Interval
+
+	return dto
+}
+
+func (r *Registration) Validate() error {
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/registration.go:61]: Done.

File: dtos/registration.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+)
+
+type Registration struct {
+	DBTimestamp   `json:",inline"`
+	ServiceId     string      `json:"serviceId"`
+	Status        string      `json:"status"`
+	Host          string      `json:"host"`
+	Port          int         `json:"port" `
+	HealthCheck   HealthCheck `json:",inline"`
+	LastConnected int64       `json:"lastConnected"`
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/registration.go:23]: Done.

File: dtos/registration.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+type HealthCheck struct {
+	Interval string `json:"interval"`
+	Path     string `json:"path"`
+	Type     string `json:"type"`
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/registration.go:29]: Done.

File: dtos/requests/registration_test.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	tests := []struct {
+		name        string
+		request     AddRegistrationRequest
+		expectedErr bool
+	}{
+		{"valid AddRegistrationRequest", valid, false},
+		{"invalid AddRegistrationRequest, empty service id", emptyServiceId, true},
+		{"invalid AddRegistrationRequest, empty port", emptyPort, true},
+		{"invalid AddRegistrationRequest, empty HealthCheck type", emptyHealthCheckType, true},
+		{"invalid AddRegistrationRequest, invalid HealthCheck interval", invalidInterval, true},
```
[the remark this answers is not in the document]
Comment (jackchenjc) [reply dtos/requests/registration_test.go:53]: Done.

File: clients/http/kvs.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply clients/http/kvs.go:27]: HI @jackchenjc, the `authInjector` can be simply added to the

File: dtos/requests/registration.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lenny-intel) [reply dtos/requests/registration.go:53]: Your correct here. Sorry, I was confused on which file this is... ;-(

**Review by lenny-intel — APPROVED** [review]
LGTM, remaining comments apply to other modules when using these additions.

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Pull request merged**